src/aws_encryption_sdk/keyrings/aws_kms/client_suppliers.py

```diff
@@ -0,0 +1,136 @@
+"""AWS KMS client suppliers for use with AWS KMS keyring.
+
+.. versionadded:: 1.5.0
+
+"""
+import logging
+
+import attr
+import six
+from attr.validators import deep_iterable, instance_of
+from botocore.client import BaseClient
+from botocore.session import Session as BotocoreSession
+
+from aws_encryption_sdk.exceptions import UnknownRegionError
+from aws_encryption_sdk.internal.validators import value_is_not_a_string
+
+from .client_cache import ClientCache
+
+try:  # Python 3.5.0 and 3.5.1 have incompatible typing modules
+    from typing import Any, Dict, Union  # noqa pylint: disable=unused-import
+except ImportError:  # pragma: no cover
+    # We only actually need these imports when running the mypy checks
+    pass
+
+_LOGGER = logging.getLogger(__name__)
+__all__ = ("ClientSupplier", "DefaultClientSupplier", "AllowRegionsClientSupplier", "DenyRegionsClientSupplier")
+
+
+class ClientSupplier(object):
+    """Base class for client suppliers.
+
+    .. versionadded:: 1.5.0
+
+    """
+
+    def client(self, region_name):
+        # type: (Union[None, str]) -> BaseClient
+        """Return a client for the requested region.
+
+        :rtype: BaseClient
+        """
+        raise NotImplementedError("'ClientSupplier' does not implement 'client'")
+
+
+@attr.s
+class DefaultClientSupplier(ClientSupplier):
+    """The default AWS KMS client supplier.
+    Creates and caches clients for any region.
+
+    .. versionadded:: 1.5.0
+
+    :param botocore_session: botocore session to use when creating clients (optional)
+    :type botocore_session: botocore.session.Session
+    """
+
+    _botocore_session = attr.ib(default=attr.Factory(BotocoreSession), validator=instance_of(BotocoreSession))
+
+    def __attrs_post_init__(self):
+        """Set up internal client cache."""
+        self._cache = ClientCache(botocore_session=self._botocore_session)
+
+    def client(self, region_name):
+        # type: (Union[None, str]) -> BaseClient
+        """Return a client for the requested region.
+
+        :rtype: BaseClient
+        """
+        return self._cache.client(region_name=region_name, service="kms")
+
+
+@attr.s
+class AllowRegionsClientSupplier(ClientSupplier):
+    """AWS KMS client supplier that only supplies clients for the specified regions.
+
+    .. versionadded:: 1.5.0
+
+    :param List[str] allowed_regions: Regions to allow
+    :param botocore_session: botocore session to use when creating clients (optional)
+    :type botocore_session: botocore.session.Session
+    """
+
+    allowed_regions = attr.ib(
+        validator=(deep_iterable(member_validator=instance_of(six.string_types)), value_is_not_a_string)
+    )
+    _botocore_session = attr.ib(default=attr.Factory(BotocoreSession), validator=instance_of(BotocoreSession))
+
+    def __attrs_post_init__(self):
+        """Set up internal client supplier."""
+        self._supplier = DefaultClientSupplier(botocore_session=self._botocore_session)
```

Comment:
Maybe nobody cares. But how do I compose _my_ client supplier with this Allow one?
Should the `_supplier` also be an option along with the `allowed_regions`?
(Or am I not reading this correctly?)

Reply:
I've been thinking about this, and I think what I'll do here is make the inner supplier an optional parameter for the filtering suppliers and the cache an optional parameter for the default supplier. That way you can plumb in whatever configurations you want by assembling the pieces.

Reply:
It is super simple to make your own.
I agree that the default should be to cache for almost all cases.